ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Memory.java

```diff
@@ -341,72 +276,26 @@ public void setBytes(
+   *     be copied from {@code value}; if {@code numBytes > value.size()}, then only the bytes in
    *     {@code value} will be copied, but the memory will be expanded if necessary to cover {@code
    *     numBytes} (in other words, {@link #getActiveWords()} will return a value consistent with
    *     having set {@code numBytes} bytes, even if less than that have been concretely set due to
    *     {@code value} being smaller).
    * @param taintedValue the bytes to copy to memory from {@code location}.
    */
   public void setBytes(final UInt256 location, final UInt256 numBytes, final Bytes taintedValue) {
-    if (numBytes.isZero()) {
-      return;
-    }
-
-    final long start = asByteIndex(location);
-    final int length = asByteLength(numBytes);
-
-    ensureCapacityForBytes(start, length);
-
-    // We've properly expanded memory as needed. We now have simply have to copy the
-    // min(length, value.size()) first bytes of value and clear any bytes that exceed value's length
-    if (taintedValue.isEmpty()) {
-      clearBytes(location, numBytes);
-      return;
-    }
-    final Bytes value;
-    if (taintedValue.size() > length) {
-      value = taintedValue.slice(0, length);
-    } else if (taintedValue.size() < length) {
-      value = taintedValue;
-      clearBytes(location.add(taintedValue.size()), numBytes.subtract(taintedValue.size()));
-    } else {
-      value = taintedValue;
-    }
-
-    // Index of last byte to set.
-    final long end = start + value.size() - 1;
-
-    final int startWord = wordForByte(start);
-    final int idxInStart = indexInWord(start);
-    final int endWord = wordForByte(end);
-
-    if (startWord == endWord) {
-      // Bytes within a word, fast-path.
-      final MutableBytes mb = data.get(startWord).mutableCopy();
-
-      value.copyTo(mb, idxInStart);
-      data.set(startWord, (Bytes32) mb.copy());
-      return;
-    }
-
-    // Spans multiple word, slower path.
-    final int bytesInStartWord = Bytes32.SIZE - idxInStart;
-
-    int valueIdx = 0;
-    final MutableBytes startMutable = data.get(startWord).mutableCopy();
-    value.slice(valueIdx, bytesInStartWord).copyTo(startMutable, idxInStart);
-    data.set(startWord, (Bytes32) startMutable.copy());
-    valueIdx += bytesInStartWord;
-    for (int i = startWord + 1; i < endWord; i++) {
-      final MutableBytes mb = data.get(i).mutableCopy();
-      value.slice(valueIdx, Bytes32.SIZE).copyTo(mb);
-      data.set(i, (Bytes32) mb.copy());
-      valueIdx += Bytes32.SIZE;
+    final int copySize = numBytes.intValue();
+    if (copySize > 0) {
+      final int srcLength = taintedValue.size();
+      final int locationInt = location.intValue();
+      ensureCapacityForBytes(locationInt, copySize);
+      if (srcLength >= copySize) {
+        System.arraycopy(taintedValue.toArrayUnsafe(), 0, data, locationInt, copySize);
+      } else {
+        Arrays.fill(data, locationInt, locationInt + copySize, (byte) 0);
```

Comment:
Leaving this comment as a marker to myself to get back to understanding why we need this line.

Reply:
For the case where what we are copying is shorter than the desired destination the EVM specifies we should fill the unspecified memory with zeros.  There are reference tests enforcing this.